src/docmanager/interfaces.ts

```diff
@@ -272,6 +272,18 @@ interface IWidgetFactory<T extends Widget> extends IDisposable {
 
 
 /**
+ * An interface for a widget extension.
+ */
+export
+interface IWidgetExtension<T extends Widget> {
+  /**
+   * Create a new extension for a given widget.
+   */
+   createNew(widget: T, model: IDocumentModel, context: IDocumentContext): IDisposable;
```

Comment:
Nice, I like how this doesn't impose on widget (i.e., to have the model or context stored on it)